File: branches/VirtualOrc/src/orc/runtime/sites/core/Not.java
//
// Not.java -- Java class Not
// Project OrcJava
//
// $Id$
//
// Copyright (c) 2009 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.runtime.sites.core;

import orc.error.runtime.TokenException;
import orc.runtime.Args;
import orc.runtime.sites.EvalSite;
import orc.type.Type;
import orc.type.structured.ArrowType;

/**
 * @author dkitchin
 */
public class Not extends EvalSite {
	@Override
	public Object evaluate(final Args args) throws TokenException {
		return !args.boolArg(0);
	}

	@Override
	public Type type() {
		return new ArrowType(Type.BOOLEAN, Type.BOOLEAN);
	}
}

File: OrcScala/src/orc/lib/bool/Not.scala
//
// Not.scala -- Scala object Not
// Project OrcScala
//
// Copyright (c) 2011 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.bool

import orc.error.runtime.{ ArgumentTypeMismatchException, ArityMismatchException }
import orc.types.{ BooleanType, SimpleFunctionType }
import orc.values.sites.{ FunctionalSite, TotalSite, TypedSite }
import orc.util.ArrayExtensions._

/** Logical negation site
  */
object Not extends TotalSite with TypedSite with FunctionalSite {
  override def name = "Not"
  def evaluate(args: Array[AnyRef]) =
    args match {
      case Array1(b: java.lang.Boolean) => new java.lang.Boolean(!b.booleanValue)
      case Array1(a) => throw new ArgumentTypeMismatchException(0, "Boolean", if (a != null) a.getClass().toString() else "null")
      case _ => throw new ArityMismatchException(1, args.size)
    }

  def orcType() = SimpleFunctionType(BooleanType, BooleanType)
}
